### PR TITLE
New Socket module

### PR DESCRIPTION
**Pull request by philpep**
This is a WIP of a new Socket module.

TODO: BSD/Darwin

### CODE REVIEW

**Comment by daenney** [comment]
Does this code actually check if something passed in as `unix://` is an actual socket? I don't see it doing so but it would be something along the lines of:

```python
import os, stat

def is_socket(path)
    return stat.S_ISSOCK(os.stat(path).st_mode))
```

**Comment by philpep** [comment]
> Does this code actually check if something passed in as unix:// is an actual socket?

Nope. The beharvior will be (assuming `/sock` doesn't exists or isn't a socket neither a listening socket)

```python
>>> Socket("unix:///sock").is_listening
False
>>> Socket("unix:///sock").clients
[]
>>> # To test if /sock is a socket (warning doesn't test if the socket is listening)
>>> File("/sock").is_socket
````

File: testinfra/modules/socket.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    elif ":" in address:
+        # tcp://127.0.0.1:22
+        # tcp://:::22
+        host, port = address.rsplit(":", 1)
+        port = int(port)
+    else:
+        # tcp://22
+        host = None
+        port = int(address)
+
```
Comment (daenney) [inline testinfra/modules/socket.py]: This still allows for `Socket("tcp://0.0.0.0)`

Comment (philpep) [inline testinfra/modules/socket.py]: Indeed. I've to check adress and port. I think this do the trick: https://stackoverflow.com/questions/319279/how-to-validate-ip-address-in-python/319298#319298

Comment (daenney) [inline testinfra/modules/socket.py]: Yes but be careful not to use `aton` but `pton` so you also support IPv6.

**Pull request merged**